File: training/go/domain/service/creature_service.go
package service

import (
	"log/slog"

	"de.sots/cellarsandcentaurs/domain/model"
)

type CreatureService struct {
	creatures Creatures
	logger    *slog.Logger
}

func NewCreatureService(creatures Creatures) *CreatureService {
	return &CreatureService{creatures: creatures, logger: slog.Default()}
}

func (s *CreatureService) Save(creature *model.Creature) error {
	s.logger.Info("saving creature", "id", creature.Id())
	return s.creatures.Save(creature)
}

func (s *CreatureService) Find(id model.CreatureId) (*model.Creature, error) {
	return s.creatures.Find(id)
}

File: sample-projects/ruby/lib/de/sots/cellars_and_centaurs/domain/service/creature_service.rb
# frozen_string_literal: true

require 'logger'

require_relative '../../version'
require_relative 'creatures'
require_relative '../model/creature'

module De::Sots::CellarsAndCentaurs::Domain::Service
  class CreatureService
    def initialize(creatures, logger = Logger.new($stdout))
      @creatures = creatures
      @logger = logger
    end

    def save(creature)
      @logger.info("saving creature #{creature.id.id}")
      @creatures.save(creature)
    end

    def find(creature_id)
      @creatures.find(creature_id)
    end
  end
end
